org.eclipse.jdt.core/model/org/eclipse/jdt/core/compiler/CompilationParticipant.java

```diff
@@ -143,6 +143,22 @@ public void processAnnotations(BuildContext[] files) {
 	// do nothing by default
 }
 
+/**
+ * Notifies this participant about a finished class compilation. This
+ * notification occurs right before the compiled class is written out to disk.
+ * This allows to perform arbitrary byte code manipulation on the generated
+ * class.
+ *
+ * @param file BuildContext of the generated class
+ * @param bytes byte representation of the generated class
+ * @return byte array with length > 0 representing the modified class or byte[0]
+ *         if no modification was applied.
+ * @since 3.34
+ */
+public byte[] postProcess(BuildContext file, byte[] bytes) {
```

Comment:
We strictly need to specify what an implementor of a CompilationParticipant can expect here if another CompilationParticipant performs post processing, too. Is there an order to be imposed? Is it deliberately random? Do we throw an exception, if two compilation participants modify the bytes? 

Please also note in the JavaDoc, how / if an implementor can / must record additional dependencies that are imposed due to the post processing. E.g. if a post processor works on the bytes for `class A` and introduces a super class reference in the bytecode to `class B`, `class A` must certainly be recompiled, if `B` changes. In that case, `file.recordDependencies` must be called, I assume. Also document, what is to be expected from UI services. E.g. would the introduced super type `B` be visible in the class hierarchy of `A`?